app/services/foreman/unattended_installation/host_verifier.rb

```diff
@@ -0,0 +1,76 @@
+module Foreman
+  module UnattendedInstallation
+    class HostVerifier
+      attr_reader :errors, :host, :request_ip, :for_host_template, :controller_name
+
+      def initialize(host, request_ip:, for_host_template:)
+        @host = host
+        @errors = []
+        @for_host_template = for_host_template
+        @request_ip = request_ip
+        @controller_name = 'unattended'
+      end
+
+      def valid?
+        return false unless valid_host_token?
+        return false unless host_found?
+        return false unless host_os?
+        return false unless host_os_family?
+
+        true
+      end
+
+      private
+
+      def valid_host_token?
+        # Incase token expires during installation
+        return true unless for_host_template
+        return true unless @host&.token_expired?
+
+        errors << {
```

Comment:
I'm not a fan of this side effect. I prefer a very simple function that only returns a boolean. Could this be moved into the `valid?` function instead? An alternative would be to raise exceptions (`HostTokenExpired`, `HostNotFound`.

Reply:
I like the exception idea. We should explore this more.

Reply:
This is intended to work similar to ActiveRecord::Validations, kind of a poor man's implementation of it 😄
see: https://guides.rubyonrails.org/active_record_validations.html#validations-overview-errors The only difference here is we return once we have a failure and only work with the one error.

I do not think that throwing exceptions is suitable for this. In rails you only want an exception when something has really gone wrong. In most cases a simple call to `valid?` doesn't throw an exception.

See how it is used in the unattended_controller: 
```ruby
. . . 
    if host_verifier.valid?
      logger.debug "Found #{@host}"
      return true
    end

    error = host_verifier.errors.first
    render_custom_error(error[:type], error[:message], error[:params])
. . .
```

Reply:
I see this as a fair implementation, I'd be more strict if this wasn't class made for this sole purpose of verification, just like in Rails. Exceptions are bad in capturing multiple errors, there is this wrapping hell we see in Java world. Some languages do not have exceptions on purpose just because of this and also because unclear contracts for runtime exceptions.

Having that said, I'd like to take the opportunity and rewrite the code in a way that we actually verify all prerequisites and report them all to the user so she can make necessary configuration changes at once.